applications/OptimizationApplication/custom_utilities/container_variable_data/expressions.cpp

```diff
@@ -0,0 +1,307 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   license: OptimizationApplication/license.txt
+//
+//  Main author:     Suneth Warnakulasuriya
+//
+
+// System includes
+#include <cmath>
+#include <vector>
+
+// Project includes
+#include "includes/define.h"
+
+// Application includes
+
+// Include base h
+#include "expressions.h"
+
+namespace Kratos {
+
+void intrusive_ptr_add_ref(const Expression* x)
+{
+    x->mReferenceCounter.fetch_add(1, std::memory_order_relaxed);
+}
+
+void intrusive_ptr_release(const Expression* x)
+{
+    if (x->mReferenceCounter.fetch_sub(1, std::memory_order_release) == 1) {
+        std::atomic_thread_fence(std::memory_order_acquire);
+        delete x;
+    }
+}
+
+LiteralDoubleExpression::LiteralDoubleExpression(const double Value)
+    : mValue(Value)
+{
+}
+
+Expression::Pointer LiteralDoubleExpression::Create(const double Value)
+{
+    return Kratos::make_intrusive<LiteralDoubleExpression>(Value);
+}
+
+double LiteralDoubleExpression::Evaluate(
+    const IndexType EntityIndex,
+    const IndexType ComponentIndex) const
+{
+    return mValue;
+}
+
+std::size_t LiteralDoubleExpression::GetDimension() const
+{
+    return 1;
+}
+
+LiteralArray3Expression::LiteralArray3Expression(
+    const array_1d<double, 3>& Value,
+    const IndexType Dimension)
+    : mValue(Value),
+      mDimension(Dimension)
+{
+}
+
+Expression::Pointer LiteralArray3Expression::Create(
+    const array_1d<double, 3>& Value,
+    const IndexType Dimension)
+{
+    return Kratos::make_intrusive<LiteralArray3Expression>(Value, Dimension);
+}
+
+double LiteralArray3Expression::Evaluate(
+    const IndexType EntityIndex,
+    const IndexType ComponentIndex) const
+{
+    return mValue[ComponentIndex];
+}
+
+std::size_t LiteralArray3Expression::GetDimension() const
+{
+    return mDimension;
+}
+
+LiteralVectorExpression::LiteralVectorExpression(
+    Kratos::shared_ptr<const Vector> pValue,
+    const IndexType Dimension)
+    : mpValue(pValue),
```

Comment:
You can avoid fiddling with the refcount unnecessarily, but I guess construction won't happen in performance critical parts.


```suggestion
    : mpValue(std::move(pValue)),
```

Reply:
Removed and rearanged nicely :D